client/beefy/src/gossip.rs

```diff
@@ -78,57 +114,25 @@ where
 	pub fn new() -> GossipValidator<B> {
 		GossipValidator {
 			topic: topic::<B>(),
-			known_votes: RwLock::new(BTreeMap::new()),
+			known_votes: RwLock::new(KnownVotes::new()),
 			next_rebroadcast: Mutex::new(Instant::now() + REBROADCAST_AFTER),
 		}
 	}
 
 	/// Note a voting round.
 	///
-	/// Noting `round` will keep `round` live.
-	///
-	/// We retain the [`MAX_LIVE_GOSSIP_ROUNDS`] most **recent** voting rounds as live.
-	/// As long as a voting round is live, it will be gossiped to peer nodes.
+	/// Noting round will start a live `round`.
 	pub(crate) fn note_round(&self, round: NumberFor<B>) {
-		debug!(target: "beefy", "🥩 About to note round #{}", round);
-
-		let mut live = self.known_votes.write();
-
-		if !live.contains_key(&round) {
-			live.insert(round, Default::default());
-		}
-
-		if live.len() > MAX_LIVE_GOSSIP_ROUNDS {
-			let to_remove = live.iter().next().map(|x| x.0).copied();
-			if let Some(first) = to_remove {
-				live.remove(&first);
-			}
-		}
+		debug!(target: "beefy", "🥩 About to note gossip round #{}", round);
+		self.known_votes.write().insert(round);
 	}
 
-	fn add_known(known_votes: &mut KnownVotes<B>, round: &NumberFor<B>, hash: MessageHash) {
-		known_votes.get_mut(round).map(|known| known.insert(hash));
-	}
-
-	// Note that we will always keep the most recent unseen round alive.
-	//
-	// This is a preliminary fix and the detailed description why we are
-	// doing this can be found as part of the issue below
-	//
-	// https://github.com/paritytech/grandpa-bridge-gadget/issues/237
-	//
-	fn is_live(known_votes: &KnownVotes<B>, round: &NumberFor<B>) -> bool {
-		let unseen_round = if let Some(max_known_round) = known_votes.keys().last() {
-			round > max_known_round
-		} else {
-			known_votes.is_empty()
-		};
-
-		known_votes.contains_key(round) || unseen_round
-	}
-
-	fn is_known(known_votes: &KnownVotes<B>, round: &NumberFor<B>, hash: &MessageHash) -> bool {
-		known_votes.get(round).map(|known| known.contains(hash)).unwrap_or(false)
+	/// Conclude a voting round.
+	///
+	/// This can be called once round is complete so we stop gossiping for it.
+	pub(crate) fn conclude_round(&self, round: NumberFor<B>) {
+		debug!(target: "beefy", "🥩 About to drop gossip round #{}", round);
+		self.known_votes.write().remove(round);
```

Comment:
Are all rounds are guaranteed to be "concluded"? There's a test below - `note_and_drop_round_works`. It "notes" rounds 1, 3, 7 and 10. And at some point 7 is concluded (this method is called). As a result - the btree map in `known_votes` will contain entries for 1, 3 and 10. If 1 and 3 are never concluded, then the map will hold entries forever => "memory leak". Please, correct me if I'm wrong